File: Algorithms/Medium/2.Add-Two-Numbers/AddTwoNumbers.py
from typing import Optional

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

# Time complexity: O(n)
# Space complexity: O(1)
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        dummy = curr = ListNode()

        add_next = 0
        while l1 or l2 or add_next:
            val_1 = 0
            if l1:
                val_1 = l1.val
                l1 = l1.next

            val_2 = 0
            if l2:
                val_2 = l2.val
                l2 = l2.next

            add_next, curr_sum = divmod(val_1 + val_2 + add_next, 10)
            curr.next = ListNode(curr_sum)
            curr = curr.next

        return dummy.next

File: Algorithms/Medium/2.Add-Two-Numbers/AddTwoNumbers.go
package medium

type ListNode struct {
	Val  int
	Next *ListNode
}

// Time complexity: O(N)
// Space complexity: O(1)
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	dummy := &ListNode{}
	curr := dummy

	addNext, val := 0, 0
	for l1 != nil || l2 != nil || addNext > 0 {
		v1, v2 := 0, 0
		if l1 != nil {
			v1 = l1.Val
			l1 = l1.Next
		}

		if l2 != nil {
			v2 = l2.Val
			l2 = l2.Next
		}

		addNext, val = divMod(v1+v2+addNext, 10)
		curr.Next = &ListNode{val, nil}
		curr = curr.Next
	}

	return dummy.Next
}

func divMod(a int, b int) (int, int) {
	quotient := a / b
	remainder := a % b

	return quotient, remainder
}
